crucible-llvm/src/Lang/Crucible/LLVM/Globals.hs

```diff
@@ -260,21 +267,57 @@ populateConstGlobals = populateGlobals f
 -- | Write the value of the given LLVMConst into the given global variable.
 --   This is intended to be used at initialization time, and will populate
 --   even read-only global data.
-populateGlobal ::
+populateGlobal :: forall sym wptr.
   ( ?lc :: TypeContext
   , 16 <= wptr
   , HasPtrWidth wptr
-  , IsSymInterface sym ) =>
+  , IsSymInterface sym
+  , HasCallStack
+  ) =>
   sym ->
-  L.Global ->
-  MemType ->
-  LLVMConst ->
+  L.Global {- ^ The global to populate -} ->
+  MemType {- ^ Type of the global -} ->
+  LLVMConst {- ^ Constant value to initialize with -} ->
+  GlobalInitializerMap ->
   MemImpl sym ->
   IO (MemImpl sym)
-populateGlobal sym gl memty cval mem =
-  do let symb = L.globalSym gl
-     let alignment = memTypeAlign (llvmDataLayout ?lc) memty
+populateGlobal sym gl memty cval giMap mem =
+  do let alignment = memTypeAlign (llvmDataLayout ?lc) memty
+
+     -- So that globals can populate and look up the globals they reference
+     -- during initialization
+     let populateRec :: HasCallStack
+                     => L.Symbol -> StateT (MemImpl sym) IO (LLVMPtr sym wptr)
+         populateRec symbol = do
+           memimpl0 <- get
+           memimpl <-
+            case Map.lookup symbol (memImplGlobalMap mem) of
+              Just _  -> pure memimpl0 -- We already populated this one
+              Nothing ->
+                -- For explanations of the various modes of failure, see the
+                -- comment on 'GlobalInitializerMap'.
+                case Map.lookup symbol giMap of
+                  Nothing -> panic "populateGlobal" $
```

Comment:
Should these error conditions actually be `panic`s? We usually reserve those for things that can only occur as a result of programming bugs, but these seem like they could be caused by malformed inputs.

Reply:
Oh, I suppose that's right. I was being generous in assuming LLVM wouldn't give us something malformed :-) What should we use? Maybe `fail`? It seems pretty unrecoverable.

Reply:
Yeah, `fail` is probably the best thing for now... I think that's what we do when we encounter type errors and such in LLVM translation.  The difference is that panic will crash all the way out of e.g. SAW, whereas a `fail` will get caught and dump you back at the repl.